README.md
# ps-tree

[![Build Status](https://travis-ci.org/indexzero/ps-tree.svg)](https://travis-ci.org/indexzero/ps-tree)
[![Code Climate](https://codeclimate.com/github/indexzero/ps-tree/badges/gpa.svg)](https://codeclimate.com/github/indexzero/ps-tree)
[![Test Coverage](https://codeclimate.com/github/indexzero/ps-tree/badges/coverage.svg)](https://codeclimate.com/github/indexzero/ps-tree)
[![npm Version](https://img.shields.io/npm/v/ps-tree.svg)](https://www.npmjs.com/package/ps-tree)
[![Node.js Version](https://img.shields.io/node/v/ps-tree.svg?style=flat)](http://nodejs.org/download/)
[![Dependency Status](https://david-dm.org/indexzero/ps-tree.svg)](https://david-dm.org/indexzero/ps-tree)

Sometimes you cannot kill child processes like you would expect, this a feature of UNIX.

>in UNIX, a process may terminate by using the exit call, and it's parent process may wait for that event by using the wait system call. the wait system call returns the process identifier of a terminated child, so that the parent tell which of the possibly many children has terminated. If the parent terminates, however, all it's children have assigned as their new parent the init process. Thus, the children still have a parent to collect their status and execution statistics.
> (from "operating system concepts")

Solution: use `ps-tree` to get all processes that a `child_process` may have started, so that they may all be terminated.

``` js
var cp = require('child_process'),
    psTree = require('ps-tree');

var child = cp.exec("node -e 'while (true);'", function () {...});

// This will not actually kill the child it will kill the `sh` process.
child.kill();
```

wtf? it's because exec actually works like this:

``` js
function exec (cmd, cb) {
  spawn('sh', ['-c', cmd]);
  ...
}
```

`sh` starts parses the command string and starts processes, and waits for them to terminate, but `exec` returns a process object with the pid of the `sh` process.
However, since it is in `wait` mode killing it does not kill the children.

Use `ps-tree` like this:

``` js
var cp = require('child_process'),
    psTree = require('ps-tree');

var child = cp.exec("node -e 'while (true);'", function () { /*...*/ });

psTree(child.pid, function (err, children) {
  cp.spawn('kill', ['-9'].concat(children.map(function (p) { return p.PID })));
});
```

If you prefer to run **psTree** from the command line, use: `node ./bin/ps-tree.js`

## Cross Platform support


The `ps-tree` module behaves differently on *nix vs. Windows by spawning different programs and parsing their output. This is based on `process.platform` and not on checking to see if a `ps` compatible program exists on the system.

#### *nix

1. " <defunct> " need to be striped
```bash
$ ps -A -o comm,ppid,pid,stat
COMMAND          PPID   PID STAT
bbsd             2899 16958 Ss
watch <defunct>  1914 16964 Z
ps              20688 16965 R+
```

### Windows
1. `wmic PROCESS WHERE ParentProcessId=4604 GET Name,ParentProcessId,ProcessId,Status)`
2. The order of head columns is fixed
```shell
> wmic PROCESS GET Name,ProcessId,ParentProcessId,Status
Name                          ParentProcessId  ProcessId   Status
System Idle Process           0                0
System                        0                4
smss.exe                      4                228
```

### Mac/Darwin

1. " " need to be striped

```shell
$ ps -A -o comm,ppid,pid,stat
COMM              PPID   PID STAT
/sbin/launchd        0     1 Ss
/usr/libexec/Use     1    43 Ss
```

### LICENSE: MIT


CHANGELOG.md
# CHANGELOG

## 1.1.1

- [#34] Locks `event-stream` to `3.3.4`.


test/test.js
var path = require('path');
var test  = require('tape');
var chalk = require('chalk');
var cp = require('child_process');
var treeKill = require('tree-kill');
var psTree = require('../');

var red = chalk.red,
    green = chalk.green,
    cyan = chalk.cyan;

var scripts = {
  parent: path.join(__dirname, 'exec', 'parent.js'),
  child: path.join(__dirname, 'exec', 'child.js')
};

test(cyan('Spawn a Parent process which has a Two Child Processes'), function (t) {
  var parent = cp.exec('node ' + scripts.parent, function (error, stdout, stderr) {});

  setTimeout(function () {
    psTree(parent.pid, function (err, children) {
      if (err) { console.log(err); }
      console.log(red('Children: '), children, '\n');
      t.true(children.length > 0, green('✓ There are ' + children.length + ' active child processes'));
      treeKill(parent.pid);
    });

    setTimeout(function () {
      psTree(parent.pid, function (err, children) {
        if (err) { console.log(err); }
        // console.log('Children: ', children, '\n');
        // console.log(' ')
        t.equal(children.length, 0, green('✓ No more active child processes (we killed them)'));
        t.end();
      });
    }, 2000); // give psTree time to kill the processes
  }, 500); // give the child process time to spawn
  // need more time on a slow(or heavy load server). maybe promise.then is better instead of the timeout
});

test(cyan('FORCE ERROR by calling psTree without supplying a Callback'), function (t) {
  var errmsg = 'Error: childrenOfPid(pid, callback) expects callback'
  // Attempt to call psTree without a callback
  try { psTree(1234); }
  catch (e) {
    t.equal(e.toString(), errmsg, green('✓ Fails when no callback supplied (as expected)'))
  }

  t.end();
});

test(cyan('Includes itself it includeRoot is true'), function (t) {
  psTree(process.pid, true, function(err, children) {
    if (err) { console.log(err); }
    t.equal(children.length, 2, green('✓ Two processes found'));

    var current;
    var other;
    if (children[0].PID === '' + process.pid) {
      current = children[0];
      other = children[1];
    } else {
      current = children[1];
      other = children[0];
    };

    t.equal(current.PID, '' + process.pid, green('✓ Current PID is valid'));
    t.equal(current.COMMAND, 'node', green('✓ Current COMM is node'));
    t.notEqual(other.PID, '' + process.pid, green('✓ Current PID is valid'));
    t.equal(other.COMMAND, 'ps', green('✓ Current COMM is ps'));
    t.end();
  });
});

test(cyan('Spawn a Child Process and psTree with a String as pid'), function (t) {
  var child = cp.exec('node ' + scripts.child, function(error, stdout, stderr) {});
  setTimeout(function(){
    psTree(child.pid.toString(), function (err, children) {
      if (err) { console.log(err); }
      // cp.spawn('kill', ['-9'].concat(children.map(function (p) { return p.PID })))
      treeKill(child.pid);
    });

    setTimeout(function() {
      psTree(child.pid.toString(), function (err, children) {
        if (err) { console.log(err); }
        t.equal(children.length, 0, green('✓ No more active child processes'));
        t.end();
      });
    }, 1000); // give psTree time to kill the processes
  }, 200); // give the child process time to spawn
});


test/direct.js
var test  = require('tape');
var chalk = require('chalk');
var cp = require('child_process');

var green = chalk.green,
    cyan = chalk.cyan;

test(cyan('Directly Execute bin/ps-tree.js'), function (t) {
  var child = cp.exec('node ./bin/ps-tree.js', function (error, data) {
    if (error !== null) {
      t.error(err);
      t.end();
      return;
    }
    t.true(child.pid, green('✓ Called ./bin/ps-tree.js directly. worked as expected'));
    t.end();
  });
});


test/exec/child.js
var started = false;
setInterval(function() {
  if (started) return;
  console.log(process.pid);
  started = true;
}, 100); // Does nothing, but prevents exit


index.js
'use strict';

var spawn = require('child_process').spawn,
    es    = require('event-stream');

module.exports = function childrenOfPid(pid, callback) {
  var headers = null;

  if (typeof callback !== 'function') {
    throw new Error('childrenOfPid(pid, callback) expects callback');
  }

  if (typeof pid === 'number') {
    pid = pid.toString();
  }

  //
  // The `ps-tree` module behaves differently on *nix vs. Windows
  // by spawning different programs and parsing their output.
  //
  // Linux:
  // 1. " <defunct> " need to be striped
  // ```bash
  // $ ps -A -o comm,ppid,pid,stat
  // COMMAND          PPID   PID STAT
  // bbsd             2899 16958 Ss
  // watch <defunct>  1914 16964 Z
  // ps              20688 16965 R+
  // ```
  //
  // Darwin:
  // $ ps -A -o comm,ppid,pid,stat
  // COMM              PPID   PID STAT
  // /sbin/launchd        0     1 Ss
  // /usr/libexec/Use     1    43 Ss
  //
  // Win32:
  // 1. wmic PROCESS WHERE ParentProcessId=4604 GET Name,ParentProcessId,ProcessId,Status)
  // 2. The order of head columns is fixed
  // ```shell
  // > wmic PROCESS GET Name,ProcessId,ParentProcessId,Status
  // Name                          ParentProcessId  ProcessId   Status
  // System Idle Process           0                0
  // System                        0                4
  // smss.exe                      4                228
  // ```

  var processLister;
  if (process.platform === 'win32') {
    // See also: https://github.com/nodejs/node-v0.x-archive/issues/2318
    processLister = spawn('wmic.exe', ['PROCESS', 'GET', 'Name,ProcessId,ParentProcessId,Status']);
  } else {
    processLister = spawn('ps', ['-A', '-o', 'ppid,pid,stat,comm']);
  }

  es.connect(
    // spawn('ps', ['-A', '-o', 'ppid,pid,stat,comm']).stdout,
    processLister.stdout,
    es.split(),
    es.map(function (line, cb) { //this could parse alot of unix command output
      var columns = line.trim().split(/\s+/);
      if (!headers) {
        headers = columns;

        //
        // Rename Win32 header name, to as same as the linux, for compatible.
        //
        headers = headers.map(normalizeHeader);
        return cb();
      }

      var row = {};
      // For each header
      var h = headers.slice();
      while (h.length) {
        row[h.shift()] = h.length ? columns.shift() : columns.join(' ');
      }

      return cb(null, row);
    }),
    es.writeArray(function (err, ps) {
      var parents = {},
          children = [];

      parents[pid] = true;
      ps.forEach(function (proc) {
        if (parents[proc.PPID]) {
          parents[proc.PID] = true;
          children.push(proc)
        }
      });

      callback(null, children);
    })
  ).on('error', callback)
}

/**
 * Normalizes the given header `str` from the Windows
 * title to the *nix title.
 *
 * @param {string} str Header string to normalize
 */
function normalizeHeader(str) {
  switch (str) {
    case 'Name':  // for win32
    case 'COMM':  // for darwin
      return 'COMMAND';
      break;
    case 'ParentProcessId':
      return 'PPID';
      break;
    case 'ProcessId':
      return 'PID';
      break;
    case 'Status':
      return 'STAT';
      break;
    default:
      return str
  }
}
